# Oakland 311 calls part 2 assignment

## Importing modules
We start by imporing the modules as follows:


In [2]:
import pandas as pd
import altair as alt


# Importing the data
Now that we have the pandas and altair modules imported, we need to import the data from the clean Oakland data we worked on last week that is oakland_311_clean.csv, rather than raw data from the internet.

As we import the data, we want to ensure the dtypes are as we want them as well as change the 'REQUESTID' from an int dtype to an object

In [3]:
oakland_311 = pd.read_csv('oakland_311_clean.csv', 
    dtype={
        'REQUESTID': object,
    },
    parse_dates=['DATETIMEINIT', 'DATETIMECLOSED', 'DATECLOSED']
)

# Let's now view the data 
This is to make sure that we have the right data with the variables in the desired dtype

In [4]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0    927708 2019-07-24 07:34:04  SeeClickFix   
1    930512 2019-08-01 13:38:16  SeeClickFix   
2    928898 2019-07-28 09:20:50  SeeClickFix   
3    930760 2019-08-02 10:13:45        Phone   
4    923520 2019-07-11 08:45:50  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO      DATETIMECLOSED          SRX  \
0                                   NaN 2019-09-30 10:25:11  6052877.252   
1                                   NaN 2019-08-04 10:38:05  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                 NaT  6081509.580   
3                                   NaN 2019-10-28 11:30:12  6079075.972   
4                                   NaN                 NaT  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  \
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA   
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA   
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA   
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA   
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA   

         DATECLOSED  
0  68 days 02:51:07  
1   2 days 20:59:49  
2               NaN  
3  87 days 01:16:27  
4               NaN

and now to view the dtypes for each variable


In [5]:
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 17 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        852541 non-null  object        
 1   DATETIMEINIT     852541 non-null  datetime64[ns]
 2   SOURCE           852541 non-null  object        
 3   DESCRIPTION      852541 non-null  object        
 4   REQCATEGORY      852460 non-null  object        
 5   REQADDRESS       836654 non-null  object        
 6   STATUS           852540 non-null  object        
 7   REFERREDTO       70168 non-null   object        
 8   DATETIMECLOSED   610875 non-null  datetime64[ns]
 9   SRX              836654 non-null  float64       
 10  SRY              836654 non-null  float64       
 11  COUNCILDISTRICT  809790 non-null  object        
 12  BEAT             812245 non-null  object        
 13  PROBADDRESS      852064 non-null  object        
 14  City             852

we can see that the REQUESTID is dtype object as we wanted but DATECLOSED did not change to dtype timedelta as DATETIMEINIT and DATETIMECLOSED. So we are going to use the command pd.to_time delta to correct this

In [6]:
oakland_311['DATECLOSED'] = pd.to_timedelta(oakland_311['DATECLOSED'])

Now let's examine the dtype in our dataframe checking for DATECLOSED as timedelta dtype

In [7]:
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 17 columns):
 #   Column           Non-Null Count   Dtype          
---  ------           --------------   -----          
 0   REQUESTID        852541 non-null  object         
 1   DATETIMEINIT     852541 non-null  datetime64[ns] 
 2   SOURCE           852541 non-null  object         
 3   DESCRIPTION      852541 non-null  object         
 4   REQCATEGORY      852460 non-null  object         
 5   REQADDRESS       836654 non-null  object         
 6   STATUS           852540 non-null  object         
 7   REFERREDTO       70168 non-null   object         
 8   DATETIMECLOSED   610875 non-null  datetime64[ns] 
 9   SRX              836654 non-null  float64        
 10  SRY              836654 non-null  float64        
 11  COUNCILDISTRICT  809790 non-null  object         
 12  BEAT             812245 non-null  object         
 13  PROBADDRESS      852064 non-null  object         
 14  City

# Let's now explore the data
What I am looking for here is to understand the type of incidents oakland residents called 311 to report. I will do this by exploring the data for the year 2020.

To do this, I will create a new dataframe with data with all the incidents reported in 2020, and to ensure that my original oakland_311.info data remains untampered with


In [10]:
oakland_311_calls_2020 = oakland_311 [
    (oakland_311['DATETIMEINIT']  >= '2020-01-01') &
    (oakland_311['DATETIMECLOSED'] < '2021-01-01')
    
].copy()

In [11]:
oakland_311_calls_2020

REQUESTID        DATETIMEINIT          SOURCE  \
4173     1060662 2020-10-28 08:21:55     SeeClickFix   
6241     1057742 2020-10-19 08:13:43           Phone   
14387    1064947 2020-11-12 15:01:32     SeeClickFix   
24528    1060772 2020-10-28 11:48:51           Phone   
31674    1064825 2020-11-12 11:13:16     SeeClickFix   
...          ...                 ...             ...   
228811   1071545 2020-12-04 13:03:45           Phone   
228813   1072173 2020-12-07 12:40:02  Phone or Email   
228817   1073951 2020-12-15 11:30:12           Phone   
228819   1075542 2020-12-22 09:37:22       Voicemail   
228821   1075885 2020-12-23 12:45:16           Phone   

                                       DESCRIPTION   REQCATEGORY  \
4173    Illegal Dumping - debris, appliances, etc.       ILLDUMP   
6241                              Tree - Limb Down         TREES   
14387   Illegal Dumping - debris, appliances, etc.       ILLDUMP   
24528         Illegal Dumping � mattress/boxspring       ILLDUMP   
31674         Illegal Dumping � mattress/boxspring       ILLDUMP   
...                                            ...           ...   
228811  Illegal Dumping - debris, appliances, etc.       ILLDUMP   
228813                 Parking - Abandoned Vehicle        POLICE   
228817                 Parking - Meter Maintenance  METER_REPAIR   
228819           Recycling Hotline - Miscellaneous     RECYCLING   
228821               Street Light - Outage/Damaged    ELECTRICAL   

                                       REQADDRESS  STATUS REFERREDTO  \
4173     (37.75323385547596, -122.19384238019262)  CLOSED        NaN   
6241     (37.73314054386291, -122.13660759117438)  CLOSED        NaN   
14387    (37.76689859111662, -122.21140629165139)  CLOSED        NaN   
24528    (37.76206734586794, -122.17270386111848)  CLOSED        NaN   
31674   (37.804301289339584, -122.27787368550469)  CLOSED        NaN   
...                                           ...     ...        ...   
228811                                        NaN  CLOSED        NaN   
228813                                        NaN  CLOSED        NaN   
228817                                        NaN  CLOSED        NaN   
228819                                        NaN  CLOSED        NaN   
228821                                        NaN  CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
4173   2020-10-30 18:00:34  6071993.226  2101162.977            CCD7  26Y   
6241   2020-10-21 13:09:41  6088409.592  2093553.319            CCD7  35Y   
14387  2020-11-16 14:40:10  6067007.217  2106230.130            CCD5  26X   
24528  2020-10-29 21:39:07  6078160.999  2104269.000            CCD6  30X   
31674  2020-11-16 19:36:35  6048056.673  2120204.961            CCD3  03Y   
...                    ...          ...          ...             ...  ...   
228811 2020-12-07 19:46:07          NaN          NaN             NaN  NaN   
228813 2020-12-07 13:07:04          NaN          NaN             NaN  NaN   
228817 2020-12-18 07:18:18          NaN          NaN             NaN  NaN   
228819 2020-12-22 09:37:31          NaN          NaN             NaN  NaN   
228821 2020-12-24 08:27:50          NaN          NaN             NaN  NaN   

                     PROBADDRESS     City State      DATECLOSED  
4173                867 75TH AVE  Oakland    CA 2 days 09:38:39  
6241               328 MARLOW DR  Oakland    CA 2 days 04:55:58  
14387              1058 50TH AVE  Oakland    CA 3 days 23:38:38  
24528   RITCHIE ST & BANCROFT AV  Oakland    CA 1 days 09:50:16  
31674                675 11TH ST  Oakland    CA 4 days 08:23:19  
...                          ...      ...   ...             ...  
228811               695 31ST ST  Oakland    CA 3 days 06:42:22  
228813          7620 GARFIELD AV  Oakland    CA 0 days 00:27:02  
228817             3219 GRAND AV  Oakland    CA 2 days 19:48:06  
228819                        ZZ  Oakland    CA 0 days 00:00:09  
2

I will then reset the index to a sequential order using the command df.reset_index(drop=True)

In [12]:
oakland_311_calls_2020 = oakland_311_calls_2020.reset_index(drop=True)

In [13]:
oakland_311_calls_2020

REQUESTID        DATETIMEINIT          SOURCE  \
0       1060662 2020-10-28 08:21:55     SeeClickFix   
1       1057742 2020-10-19 08:13:43           Phone   
2       1064947 2020-11-12 15:01:32     SeeClickFix   
3       1060772 2020-10-28 11:48:51           Phone   
4       1064825 2020-11-12 11:13:16     SeeClickFix   
...         ...                 ...             ...   
70769   1071545 2020-12-04 13:03:45           Phone   
70770   1072173 2020-12-07 12:40:02  Phone or Email   
70771   1073951 2020-12-15 11:30:12           Phone   
70772   1075542 2020-12-22 09:37:22       Voicemail   
70773   1075885 2020-12-23 12:45:16           Phone   

                                      DESCRIPTION   REQCATEGORY  \
0      Illegal Dumping - debris, appliances, etc.       ILLDUMP   
1                                Tree - Limb Down         TREES   
2      Illegal Dumping - debris, appliances, etc.       ILLDUMP   
3            Illegal Dumping � mattress/boxspring       ILLDUMP   
4            Illegal Dumping � mattress/boxspring       ILLDUMP   
...                                           ...           ...   
70769  Illegal Dumping - debris, appliances, etc.       ILLDUMP   
70770                 Parking - Abandoned Vehicle        POLICE   
70771                 Parking - Meter Maintenance  METER_REPAIR   
70772           Recycling Hotline - Miscellaneous     RECYCLING   
70773               Street Light - Outage/Damaged    ELECTRICAL   

                                      REQADDRESS  STATUS REFERREDTO  \
0       (37.75323385547596, -122.19384238019262)  CLOSED        NaN   
1       (37.73314054386291, -122.13660759117438)  CLOSED        NaN   
2       (37.76689859111662, -122.21140629165139)  CLOSED        NaN   
3       (37.76206734586794, -122.17270386111848)  CLOSED        NaN   
4      (37.804301289339584, -122.27787368550469)  CLOSED        NaN   
...                                          ...     ...        ...   
70769                                        NaN  CLOSED        NaN   
70770                                        NaN  CLOSED        NaN   
70771                                        NaN  CLOSED        NaN   
70772                                        NaN  CLOSED        NaN   
70773                                        NaN  CLOSED        NaN   

           DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0     2020-10-30 18:00:34  6071993.226  2101162.977            CCD7  26Y   
1     2020-10-21 13:09:41  6088409.592  2093553.319            CCD7  35Y   
2     2020-11-16 14:40:10  6067007.217  2106230.130            CCD5  26X   
3     2020-10-29 21:39:07  6078160.999  2104269.000            CCD6  30X   
4     2020-11-16 19:36:35  6048056.673  2120204.961            CCD3  03Y   
...                   ...          ...          ...             ...  ...   
70769 2020-12-07 19:46:07          NaN          NaN             NaN  NaN   
70770 2020-12-07 13:07:04          NaN          NaN             NaN  NaN   
70771 2020-12-18 07:18:18          NaN          NaN             NaN  NaN   
70772 2020-12-22 09:37:31          NaN          NaN             NaN  NaN   
70773 2020-12-24 08:27:50          NaN          NaN             NaN  NaN   

                    PROBADDRESS     City State      DATECLOSED  
0                  867 75TH AVE  Oakland    CA 2 days 09:38:39  
1                 328 MARLOW DR  Oakland    CA 2 days 04:55:58  
2                 1058 50TH AVE  Oakland    CA 3 days 23:38:38  
3      RITCHIE ST & BANCROFT AV  Oakland    CA 1 days 09:50:16  
4                   675 11TH ST  Oakland    CA 4 days 08:23:19  
...                         ...      ...   ...             ...  
70769               695 31ST ST  Oakland    CA 3 days 06:42:22  
70770          7620 GARFIELD AV  Oakland    CA 0 days 00:27:02  
70771             3219 GRAND AV  Oakland    CA 2 days 19:48:06  
70772                        ZZ  Oakland    CA 0 days 00:00:09  
70773            4120 WALNUT ST  Oakland    CA 0 days 19:42

# Now let's explore REQCATEGORY and see what the 311 requests were about in 2020
To do this we will use the series.unique() command on the REQCATEGORY column

In [15]:
oakland_311_calls_2020['REQCATEGORY'].unique()

array(['ILLDUMP', 'TREES', 'POLICE', 'BLDGMAINT', 'METER_REPAIR',
       'DRAINAGE', 'GRAFFITI', 'PARKING', 'ELECTRICAL', 'OTHER',
       'STREETSW', 'RECYCLING', 'TRAFFIC', 'CUT_CLEAN', 'SEWERS', 'PARKS',
       'KOCB', 'ROW_INSPECTORS', 'HE_CLEAN', 'TRAFFIC_ENGIN',
       'ROW_STREETSW', 'CW_DIT_GIS', 'WATERSHED', 'FIRE', 'ENVIRON_ENF'],
      dtype=object)

We can see that the calls were categoriesed into such categories as 'ILLDUMP', 'PARKING', 'FIRE' and more. But we need to know just how many calls under these categories were received. To do so, we will use the command df.groupy() and df.count()


In [16]:
oakland_311_calls_2020.groupby(['REQCATEGORY']).count()

REQUESTID  DATETIMEINIT  SOURCE  DESCRIPTION  REQADDRESS  \
REQCATEGORY                                                                
BLDGMAINT            3794          3794    3794         3794        3760   
CUT_CLEAN             317           317     317          317         316   
CW_DIT_GIS             32            32      32           32          19   
DRAINAGE              618           618     618          618         604   
ELECTRICAL           1842          1842    1842         1842        1823   
ENVIRON_ENF            45            45      45           45          44   
FIRE                   11            11      11           11          11   
GRAFFITI             3287          3287    3287         3287        3284   
HE_CLEAN               53            53      53           53          51   
ILLDUMP             35056         35056   35056        35056       34873   
KOCB                  407           407     407          407         401   
METER_REPAIR          465           465     465          465         451   
OTHER                2043          2043    2043         2043        2032   
PARKING               868           868     868          868         863   
PARKS                 992           992     992          992         981   
POLICE              14287         14287   14287        14287       14005   
RECYCLING            1664          1664    1664         1664        1360   
ROW_INSPECTORS         70            70      70           70          70   
ROW_STREETSW           38            38      38           38          36   
SEWERS                832           832     832          832         827   
STREETSW             1038          1038    1038         1038        1035   
TRAFFIC              1571          1571    1571         1571        1516   
TRAFFIC_ENGIN         126           126     126          126         120   
TREES                1295          1295    1295         1295        1270   
WATERSHED              23            23      23           23          19   

                STATUS  REFERREDTO  DATETIMECLOSED    SRX    SRY  \
REQCATEGORY                                                        
BLDGMAINT         3794           0            3794   3760   3760   
CUT_CLEAN          317           4             317    316    316   
CW_DIT_GIS          32           0              32     19     19   
DRAINAGE           618           0             618    604    604   
ELECTRICAL        1842           1            1842   1823   1823   
ENVIRON_ENF         45           1              45     44     44   
FIRE                11           0              11     11     11   
GRAFFITI          3287          78            3287   3284   3284   
HE_CLEAN            53           0              53     51     51   
ILLDUMP          35056           1           35056  34873  34873   
KOCB               407           0             407    401    401   
METER_REPAIR       465           0             465    451    451   
OTHER             2043         147            2043   2032   2032   
PARKING            868           0             868    863    863   
PARKS              992           0             992    981    981   
POLICE           14287           1           14287  14005  14005   
RECYCLING         1664           2            1664   1360   1360   
ROW_INSPECTORS      70          10              70     70     70   
ROW_STREETSW        38           0              38     36     36   
SEWERS             832           5             832    827    827   
STREETSW          1038           2            1038   1035   1035   
TRAFFIC           1571           0            1571   1516   1516   
TRAFFIC_ENGIN      126           0             126    120    120   
TREES             1295           0            1295   1270   1270   
WATERSHED           23           2              23     19     19   

                COUNCILDISTRICT   BEAT  PROBADDRESS   City  State  DATECLOSED  
REQCATEGORY                                   

We now need to set the results of the groupby/count to a new dataframe called category_counts_2020....

In [17]:
category_counts_2020 = oakland_311_calls_2020.groupby(['REQCATEGORY']).count()

...and then subset it by REQUESTID

In [18]:
category_counts_2020 = category_counts_2020[['REQUESTID']].copy()

# Let us now view category_counts_2020 

In [19]:
category_counts_2020

REQUESTID
REQCATEGORY              
BLDGMAINT            3794
CUT_CLEAN             317
CW_DIT_GIS             32
DRAINAGE              618
ELECTRICAL           1842
ENVIRON_ENF            45
FIRE                   11
GRAFFITI             3287
HE_CLEAN               53
ILLDUMP             35056
KOCB                  407
METER_REPAIR          465
OTHER                2043
PARKING               868
PARKS                 992
POLICE              14287
RECYCLING            1664
ROW_INSPECTORS         70
ROW_STREETSW           38
SEWERS                832
STREETSW             1038
TRAFFIC              1571
TRAFFIC_ENGIN         126
TREES                1295
WATERSHED              23

# Resetting the index
As we can see, REQCATEGORY is not a column but an index...

In [20]:
category_counts_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Case_Count  25 non-null     object
 1   Case_Count  25 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 528.0+ bytes


# ...we need to correct thus by making REQCATEGORY a column and introducing an index
This we can do using df.reset_index(). We need an index, so we're not going to use the drop=True

In [120]:
category_counts_2020 = category_counts_2020.reset_index()

AttributeError: 'NoneType' object has no attribute 'reset_index'

# Now let's view the dataframe

In [22]:
category_counts_2020

# Let's now rename our columns 
We want to rename REQUESTID to CASE_COUNT. We will use the list way (df.columns) because we only have two columns in this dataframe

In [ ]:
category_counts_2020.columns = ['REQUESTID', 'Case_Count']

AttributeError: 'NoneType' object has no attribute 'rename'

# Let's view our new columns now

In [ ]:
category_counts_2020

# Sorting the data
Now let's sort the data so we can see which Case-Count received more calls in 2020. To do this, we will set by descending order
Pandas sorts by ascending order, so we will use ascending=False

In [25]:
category_counts_2020 = category_counts_2020.sort_values(by=['Case_Count'], ascending=False).copy()



# Let's take a look at the sorted data..


In [26]:
category_counts_2020

REQUESTID  Case_Count
9          ILLDUMP       35056
15          POLICE       14287
0        BLDGMAINT        3794
7         GRAFFITI        3287
12           OTHER        2043
4       ELECTRICAL        1842
16       RECYCLING        1664
21         TRAFFIC        1571
23           TREES        1295
20        STREETSW        1038
14           PARKS         992
13         PARKING         868
19          SEWERS         832
3         DRAINAGE         618
11    METER_REPAIR         465
10            KOCB         407
1        CUT_CLEAN         317
22   TRAFFIC_ENGIN         126
17  ROW_INSPECTORS          70
8         HE_CLEAN          53
5      ENVIRON_ENF          45
18    ROW_STREETSW          38
2       CW_DIT_GIS          32
24       WATERSHED          23
6             FIRE          11

ILLDUMP, which I presume is illegal dumping, received 35,056 calls in 2020 followed by the Police with 14,287. But we can see that our index is jumbled up, so lets correct this with reset.index(drop=True)

In [27]:
category_counts_2020 = category_counts_2020.reset_index(drop=True)

And let's view the data again...

In [28]:
category_counts_2020

REQUESTID  Case_Count
0          ILLDUMP       35056
1           POLICE       14287
2        BLDGMAINT        3794
3         GRAFFITI        3287
4            OTHER        2043
5       ELECTRICAL        1842
6        RECYCLING        1664
7          TRAFFIC        1571
8            TREES        1295
9         STREETSW        1038
10           PARKS         992
11         PARKING         868
12          SEWERS         832
13        DRAINAGE         618
14    METER_REPAIR         465
15            KOCB         407
16       CUT_CLEAN         317
17   TRAFFIC_ENGIN         126
18  ROW_INSPECTORS          70
19        HE_CLEAN          53
20     ENVIRON_ENF          45
21    ROW_STREETSW          38
22      CW_DIT_GIS          32
23       WATERSHED          23
24            FIRE          11

# Let's now visualize this data using altair 
We are going to make a bar chart using the data



In [121]:
alt.Chart(category_counts_2020).mark_bar().encode(
    x='Case_Count',
    y='REQUESTID'
).properties(
    title='OAKLAND 311 CASES: 2020'
)

/Users/kennethkipruto/.pyenv/versions/3.8.5/envs/lecture_06-3.8.5/lib/python3.8/site-packages/altair/vegalite/v4/api.py:96: UserWarning: data of type <class 'NoneType'> not recognized
  warnings.warn("data of type {} not recognized".format(type(data)))


ValueError: Case_Count encoding field is specified without a type; the type cannot be automatically inferred because the data is not specified as a pandas.DataFrame.

alt.Chart(...)

# The resulting chart is jumbled up,
and we need to rearrange it to descending order by inverting the x-axis in in Y

In [32]:
alt.Chart(category_counts_2020).mark_bar().encode(
    x='Case_Count',
    y=alt.Y('REQUESTID', sort='-x')
).properties(
    title='OAKLAND 311 CASES: 2020'
)

alt.Chart(...)

# Now let's go back and do a count for all cases by year

We will go back to the oakland_311 calls. We do know from last week that the calls from 2009 and 2021 are not complete. We can see by using df.min() and df.max(). 

In [35]:
oakland_311['DATETIMEINIT'].min ()

Timestamp('2009-07-01 08:05:36')

In [36]:
oakland_311['DATETIMECLOSED'].max ()

Timestamp('2021-10-29 16:24:38')

# We have to subset this to only get those of complete years by subsetting
We will do this by creating a new dataframe called oakland_complete_years that has data from 2010-01-01 to 2020-12-31

In [38]:
oakland_complete_years = oakland_311[
    (oakland_311['DATETIMEINIT'] >= '2010-01-01') &
    (oakland_311['DATETIMECLOSED'] < '2021-01-01')
].reset_index(drop=True)

# Let's view oakland_complete-years

In [39]:
oakland_complete_years

REQUESTID        DATETIMEINIT       SOURCE  \
0         927708 2019-07-24 07:34:04  SeeClickFix   
1         930512 2019-08-01 13:38:16  SeeClickFix   
2         930760 2019-08-02 10:13:45        Phone   
3         917609 2019-06-21 15:54:26        Phone   
4         918548 2019-06-25 10:05:21        Email   
...          ...                 ...          ...   
540849    927445 2019-07-23 11:42:38  SeeClickFix   
540850    930717 2019-08-02 08:58:15  SeeClickFix   
540851    928953 2019-07-28 16:50:48  SeeClickFix   
540852    920576 2019-07-01 16:27:51        Phone   
540853    919254 2019-06-27 10:24:31  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
0                      Parking - Abandoned Vehicle      POLICE   
1             Illegal Dumping � mattress/boxspring     ILLDUMP   
2                     City Bldg - Structural Issue   BLDGMAINT   
3             Illegal Dumping � mattress/boxspring     ILLDUMP   
4                            Trash - Missed Pickup   RECYCLING   
...                                            ...         ...   
540849               City Bldg - Clean / Custodial   BLDGMAINT   
540850                 Parking - Abandoned Vehicle      POLICE   
540851                 Parking - Abandoned Vehicle      POLICE   
540852  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
540853               Street Light - Outage/Damaged  ELECTRICAL   

                                      REQADDRESS  STATUS REFERREDTO  \
0       (37.83332450468423, -122.26187389355277)  CLOSED        NaN   
1       (37.80947700189105, -122.29480799834606)  CLOSED        NaN   
2       (37.74458491831174, -122.16914534319535)  CLOSED        NaN   
3       (37.75806464022221, -122.19285964422009)  CLOSED        NaN   
4         (37.807414514252386, -122.28391068885)  CLOSED        NaN   
...                                          ...     ...        ...   
540849  (37.80606600187486, -122.27243850026787)  CLOSED        NaN   
540850  (37.81960600132651, -122.27702919438389)  CLOSED        NaN   
540851   (37.75789339999198, -122.1766678975261)  CLOSED        NaN   
540852   (37.7697750419514, -122.18324682258614)  CLOSED        NaN   
540853   (37.84077618000984, -122.2688993400894)  CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0      2019-09-30 10:25:11  6052877.252  2130683.728            CCD1  12X   
1      2019-08-04 10:38:05  6043201.257  2122182.593            CCD3  02Y   
2      2019-10-28 11:30:12  6079075.972  2097885.948            CCD7  32X   
3      2019-06-25 20:18:34  6072309.111  2102916.534            CCD7  26Y   
4      2019-06-26 14:19:37  6046334.461  2121371.541            CCD3  02X   
...                    ...          ...          ...             ...  ...   
540849 2019-09-18 12:33:10  6049638.825  2120817.635            CCD3  04X   
540850 2019-09-03 09:14:03  6048406.421  2125772.131            CCD3  07X   
540851 2019-12-26 13:18:26  6076988.064  2102769.930            CCD6  30X   
540852 2019-07-01 20:08:55  6075164.293  2107129.719            CCD6  30X   
540853 2019-07-05 14:57:56  6050899.986  2133434.859            CCD1  11X   

                    PROBADDRESS     City State        DATECLOSED  
0                   479 45TH ST  Oakland    CA  68 days 02:51:07  
1               1117 CHESTER ST  Oakland    CA   2 days 20:59:49  
2               HOLLY MINI PARK  Oakland    CA  87 days 01:16:27  
3                  1071 70TH AV  Oakland    CA   4 days 04:24:08  
4                  1015 12TH ST  Oakland    CA   1 days 04:14:16  
...                         ...      ...   ...               ...  
540849    250 FRANK H OGAWA PLZ  Oakland    CA  57 days 00:50:32  
540850           2825 MARKET ST  Oakland    CA  32 days 00:15:48  
540851         7924 plymouth st  Oakland    CA 150 days 20:27:38  
540852  66TH AVE & BANCROFT AVE  Oakland    CA   0 days 03:41:04  
540853              764 56TH ST  Oakland    CA   8 days 04:33:

# Now lets groupby()
This time, we will group by datetimes using pd.Grouper(key='DATETIMEINIT', axis=0, freq='A'). We need to create a new document called annual_cases for this

In [46]:
annual_cases = oakland_complete_years.groupby([pd.Grouper(key='DATETIMEINIT', axis=0, freq='A')]).count()

Let's now view the new dataframe

In [47]:
annual_cases

REQUESTID  SOURCE  DESCRIPTION  REQCATEGORY  REQADDRESS  STATUS  \
DATETIMEINIT                                                                    
2010-12-31        31504   31504        31504        31504       30584   31504   
2011-12-31        33915   33915        33915        33911       32969   33915   
2012-12-31        36560   36560        36560        36557       35607   36560   
2013-12-31        40243   40243        40243        40243       38485   40243   
2014-12-31        45399   45399        45399        45399       44406   45399   
2015-12-31        48990   48990        48990        48975       47746   48990   
2016-12-31        53969   53969        53969        53963       52454   53969   
2017-12-31        52741   52741        52741        52741       51486   52741   
2018-12-31        49873   49873        49873        49868       49161   49873   
2019-12-31        76886   76886        76886        76883       75840   76886   
2020-12-31        70774   70774        70774        70774       69771   70774   

              REFERREDTO  DATETIMECLOSED    SRX    SRY  COUNCILDISTRICT  \
DATETIMEINIT                                                              
2010-12-31             6           31504  30584  30584            28714   
2011-12-31            13           33915  32969  32969            31425   
2012-12-31           131           36560  35607  35607            35366   
2013-12-31            85           40243  38485  38485            38336   
2014-12-31           152           45399  44406  44406            43974   
2015-12-31           114           48990  47746  47746            47280   
2016-12-31           128           53969  52454  52454            51950   
2017-12-31           215           52741  51486  51486            50989   
2018-12-31            79           49873  49161  49161            48905   
2019-12-31           179           76886  75840  75840            74849   
2020-12-31           254           70774  69771  69771            69466   

               BEAT  PROBADDRESS   City  State  DATECLOSED  
DATETIMEINIT                                                
2010-12-31    28606        31421  31504  31504       31504  
2011-12-31    31471        33850  33915  33915       33915  
2012-12-31    35528        36500  36560  36560       36560  
2013-12-31    38421        40234  40243  40243       40243  
2014-12-31    44094        45391  45399  45399       45399  
2015-12-31    47409        48972  48990  48990       48990  
2016-12-31    52052        53958  53969  53969       53969  
2017-12-31    51121        52740  52741  52741       52741  
2018-12-31    49060        49861  49873  49873       49873  
2019-12-31    75171        76875  76886  76886       76886  
2020-12-31    69748        70771  70774  70774       70774

# Let us now subset the REQUESTID and reset inex

In [48]:
annual_cases = annual_cases[['REQUESTID']].reset_index()

In [49]:
annual_cases

DATETIMEINIT  REQUESTID
0    2010-12-31      31504
1    2011-12-31      33915
2    2012-12-31      36560
3    2013-12-31      40243
4    2014-12-31      45399
5    2015-12-31      48990
6    2016-12-31      53969
7    2017-12-31      52741
8    2018-12-31      49873
9    2019-12-31      76886
10   2020-12-31      70774

# And now we rename REQUESTID as CASE_Counts

In [50]:
annual_cases.columns = ['REQUESTID', 'Case_Count']

In [51]:
annual_cases

REQUESTID  Case_Count
0  2010-12-31       31504
1  2011-12-31       33915
2  2012-12-31       36560
3  2013-12-31       40243
4  2014-12-31       45399
5  2015-12-31       48990
6  2016-12-31       53969
7  2017-12-31       52741
8  2018-12-31       49873
9  2019-12-31       76886
10 2020-12-31       70774

In [85]:
annual_cases.info

<bound method DataFrame.info of     REQUESTID  Case_Count
0  2010-12-31       31504
1  2011-12-31       33915
2  2012-12-31       36560
3  2013-12-31       40243
4  2014-12-31       45399
5  2015-12-31       48990
6  2016-12-31       53969
7  2017-12-31       52741
8  2018-12-31       49873
9  2019-12-31       76886
10 2020-12-31       70774>

# And now we create a new column called year to correspond with the data
we will use the command dt.year

In [86]:
annual_cases['Year'] = annual_cases['REQUESTID'].dt.year

In [87]:
annual_cases

REQUESTID  Case_Count  Year
0  2010-12-31       31504  2010
1  2011-12-31       33915  2011
2  2012-12-31       36560  2012
3  2013-12-31       40243  2013
4  2014-12-31       45399  2014
5  2015-12-31       48990  2015
6  2016-12-31       53969  2016
7  2017-12-31       52741  2017
8  2018-12-31       49873  2018
9  2019-12-31       76886  2019
10 2020-12-31       70774  2020

# At this point, I will drop REQUESTID to remain with just two columns


In [88]:
annual_cases = annual_cases [['Case_Count', 'Year']].copy()

### Let's view the new dataframe

In [89]:
annual_cases

Case_Count  Year
0        31504  2010
1        33915  2011
2        36560  2012
3        40243  2013
4        45399  2014
5        48990  2015
6        53969  2016
7        52741  2017
8        49873  2018
9        76886  2019
10       70774  2020

## and let's visualize this new dataframe

In [90]:
alt.Chart(annual_cases).mark_bar().encode(
    x='Year:O',
    y='Case_Count'
).properties(
    title='Oakland 311 calls in 2020: Number of cases'
)

alt.Chart(...)

# Let's calculate the median
To do this, I'll first create a new column called Close_Time_Seconds that converts Close_Time to seconds. This will enable us to calculate the median number of seconds.

In [93]:
oakland_complete_years

REQUESTID        DATETIMEINIT       SOURCE  \
0         927708 2019-07-24 07:34:04  SeeClickFix   
1         930512 2019-08-01 13:38:16  SeeClickFix   
2         930760 2019-08-02 10:13:45        Phone   
3         917609 2019-06-21 15:54:26        Phone   
4         918548 2019-06-25 10:05:21        Email   
...          ...                 ...          ...   
540849    927445 2019-07-23 11:42:38  SeeClickFix   
540850    930717 2019-08-02 08:58:15  SeeClickFix   
540851    928953 2019-07-28 16:50:48  SeeClickFix   
540852    920576 2019-07-01 16:27:51        Phone   
540853    919254 2019-06-27 10:24:31  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
0                      Parking - Abandoned Vehicle      POLICE   
1             Illegal Dumping � mattress/boxspring     ILLDUMP   
2                     City Bldg - Structural Issue   BLDGMAINT   
3             Illegal Dumping � mattress/boxspring     ILLDUMP   
4                            Trash - Missed Pickup   RECYCLING   
...                                            ...         ...   
540849               City Bldg - Clean / Custodial   BLDGMAINT   
540850                 Parking - Abandoned Vehicle      POLICE   
540851                 Parking - Abandoned Vehicle      POLICE   
540852  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
540853               Street Light - Outage/Damaged  ELECTRICAL   

                                      REQADDRESS  STATUS REFERREDTO  \
0       (37.83332450468423, -122.26187389355277)  CLOSED        NaN   
1       (37.80947700189105, -122.29480799834606)  CLOSED        NaN   
2       (37.74458491831174, -122.16914534319535)  CLOSED        NaN   
3       (37.75806464022221, -122.19285964422009)  CLOSED        NaN   
4         (37.807414514252386, -122.28391068885)  CLOSED        NaN   
...                                          ...     ...        ...   
540849  (37.80606600187486, -122.27243850026787)  CLOSED        NaN   
540850  (37.81960600132651, -122.27702919438389)  CLOSED        NaN   
540851   (37.75789339999198, -122.1766678975261)  CLOSED        NaN   
540852   (37.7697750419514, -122.18324682258614)  CLOSED        NaN   
540853   (37.84077618000984, -122.2688993400894)  CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0      2019-09-30 10:25:11  6052877.252  2130683.728            CCD1  12X   
1      2019-08-04 10:38:05  6043201.257  2122182.593            CCD3  02Y   
2      2019-10-28 11:30:12  6079075.972  2097885.948            CCD7  32X   
3      2019-06-25 20:18:34  6072309.111  2102916.534            CCD7  26Y   
4      2019-06-26 14:19:37  6046334.461  2121371.541            CCD3  02X   
...                    ...          ...          ...             ...  ...   
540849 2019-09-18 12:33:10  6049638.825  2120817.635            CCD3  04X   
540850 2019-09-03 09:14:03  6048406.421  2125772.131            CCD3  07X   
540851 2019-12-26 13:18:26  6076988.064  2102769.930            CCD6  30X   
540852 2019-07-01 20:08:55  6075164.293  2107129.719            CCD6  30X   
540853 2019-07-05 14:57:56  6050899.986  2133434.859            CCD1  11X   

                    PROBADDRESS     City State        DATECLOSED  
0                   479 45TH ST  Oakland    CA  68 days 02:51:07  
1               1117 CHESTER ST  Oakland    CA   2 days 20:59:49  
2               HOLLY MINI PARK  Oakland    CA  87 days 01:16:27  
3                  1071 70TH AV  Oakland    CA   4 days 04:24:08  
4                  1015 12TH ST  Oakland    CA   1 days 04:14:16  
...                         ...      ...   ...               ...  
540849    250 FRANK H OGAWA PLZ  Oakland    CA  57 days 00:50:32  
540850           2825 MARKET ST  Oakland    CA  32 days 00:15:48  
540851         7924 plymouth st  Oakland    CA 150 days 20:27:38  
540852  66TH AVE & BANCROFT AVE  Oakland    CA   0 days 03:41:04  
540853              764 56TH ST  Oakland    CA   8 days 04:33:

In [103]:
oakland_complete_years['Close_Time_Seconds'] = oakland_complete_years['DATECLOSED'].dt.total_seconds()

### Let's view this dataframe and ensure it is ok

In [104]:
oakland_complete_years

REQUESTID        DATETIMEINIT       SOURCE  \
0         927708 2019-07-24 07:34:04  SeeClickFix   
1         930512 2019-08-01 13:38:16  SeeClickFix   
2         930760 2019-08-02 10:13:45        Phone   
3         917609 2019-06-21 15:54:26        Phone   
4         918548 2019-06-25 10:05:21        Email   
...          ...                 ...          ...   
540849    927445 2019-07-23 11:42:38  SeeClickFix   
540850    930717 2019-08-02 08:58:15  SeeClickFix   
540851    928953 2019-07-28 16:50:48  SeeClickFix   
540852    920576 2019-07-01 16:27:51        Phone   
540853    919254 2019-06-27 10:24:31  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
0                      Parking - Abandoned Vehicle      POLICE   
1             Illegal Dumping � mattress/boxspring     ILLDUMP   
2                     City Bldg - Structural Issue   BLDGMAINT   
3             Illegal Dumping � mattress/boxspring     ILLDUMP   
4                            Trash - Missed Pickup   RECYCLING   
...                                            ...         ...   
540849               City Bldg - Clean / Custodial   BLDGMAINT   
540850                 Parking - Abandoned Vehicle      POLICE   
540851                 Parking - Abandoned Vehicle      POLICE   
540852  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
540853               Street Light - Outage/Damaged  ELECTRICAL   

                                      REQADDRESS  STATUS REFERREDTO  \
0       (37.83332450468423, -122.26187389355277)  CLOSED        NaN   
1       (37.80947700189105, -122.29480799834606)  CLOSED        NaN   
2       (37.74458491831174, -122.16914534319535)  CLOSED        NaN   
3       (37.75806464022221, -122.19285964422009)  CLOSED        NaN   
4         (37.807414514252386, -122.28391068885)  CLOSED        NaN   
...                                          ...     ...        ...   
540849  (37.80606600187486, -122.27243850026787)  CLOSED        NaN   
540850  (37.81960600132651, -122.27702919438389)  CLOSED        NaN   
540851   (37.75789339999198, -122.1766678975261)  CLOSED        NaN   
540852   (37.7697750419514, -122.18324682258614)  CLOSED        NaN   
540853   (37.84077618000984, -122.2688993400894)  CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0      2019-09-30 10:25:11  6052877.252  2130683.728            CCD1  12X   
1      2019-08-04 10:38:05  6043201.257  2122182.593            CCD3  02Y   
2      2019-10-28 11:30:12  6079075.972  2097885.948            CCD7  32X   
3      2019-06-25 20:18:34  6072309.111  2102916.534            CCD7  26Y   
4      2019-06-26 14:19:37  6046334.461  2121371.541            CCD3  02X   
...                    ...          ...          ...             ...  ...   
540849 2019-09-18 12:33:10  6049638.825  2120817.635            CCD3  04X   
540850 2019-09-03 09:14:03  6048406.421  2125772.131            CCD3  07X   
540851 2019-12-26 13:18:26  6076988.064  2102769.930            CCD6  30X   
540852 2019-07-01 20:08:55  6075164.293  2107129.719            CCD6  30X   
540853 2019-07-05 14:57:56  6050899.986  2133434.859            CCD1  11X   

                    PROBADDRESS     City State        DATECLOSED  \
0                   479 45TH ST  Oakland    CA  68 days 02:51:07   
1               1117 CHESTER ST  Oakland    CA   2 days 20:59:49   
2               HOLLY MINI PARK  Oakland    CA  87 days 01:16:27   
3                  1071 70TH AV  Oakland    CA   4 days 04:24:08   
4                  1015 12TH ST  Oakland    CA   1 days 04:14:16   
...                         ...      ...   ...               ...   
540849    250 FRANK H OGAWA PLZ  Oakland    CA  57 days 00:50:32   
540850           2825 MARKET ST  Oakland    CA  32 days 00:15:48   
540851         7924 plymouth st  Oakland    CA 150 days 20:27:38   
540852  66TH AVE & BANCROFT AVE  Oakland    CA   0 days 03:41:04   
540853              764 56TH ST  Oakland    CA   8 

### Now let us calculate the median

In [106]:
median_close_time = oakland_complete_years.groupby([pd.Grouper(key='DATETIMEINIT', axis=0, freq='A'), ]).median()

In [107]:
median_close_time

SRX           SRY  DATECLOSED_Seconds  \
DATETIMEINIT                                                   
2010-12-31    6.060786e+06  2.116971e+06            271677.0   
2011-12-31    6.060636e+06  2.117948e+06            344876.0   
2012-12-31    6.060666e+06  2.118715e+06            438460.0   
2013-12-31    6.059985e+06  2.118508e+06            518555.0   
2014-12-31    6.057350e+06  2.119016e+06            428056.0   
2015-12-31    6.057153e+06  2.118799e+06            445098.0   
2016-12-31    6.060620e+06  2.117480e+06            518028.0   
2017-12-31    6.062140e+06  2.117133e+06            357021.0   
2018-12-31    6.060729e+06  2.117718e+06            352086.0   
2019-12-31    6.062151e+06  2.116865e+06            511923.0   
2020-12-31    6.062948e+06  2.115295e+06            360122.5   

              Close_Time_Seconds  
DATETIMEINIT                      
2010-12-31              271677.0  
2011-12-31              344876.0  
2012-12-31              438460.0  
2013-12-31              518555.0  
2014-12-31              428056.0  
2015-12-31              445098.0  
2016-12-31              518028.0  
2017-12-31              357021.0  
2018-12-31              352086.0  
2019-12-31              511923.0  
2020-12-31              360122.5

### We want DATETIMEINIT on its own column
We will use the reset_index() to do this

In [108]:
median_close_time = median_close_time[['Close_Time_Seconds']].reset_index()

In [109]:
median_close_time


DATETIMEINIT  Close_Time_Seconds
0    2010-12-31            271677.0
1    2011-12-31            344876.0
2    2012-12-31            438460.0
3    2013-12-31            518555.0
4    2014-12-31            428056.0
5    2015-12-31            445098.0
6    2016-12-31            518028.0
7    2017-12-31            357021.0
8    2018-12-31            352086.0
9    2019-12-31            511923.0
10   2020-12-31            360122.5

## I will then create a new column called 'Year' and convert seconds to days

In [110]:
median_close_time['Year'] = median_close_time['DATETIMEINIT'].dt.year
median_close_time['Close_Time_Days'] = median_close_time['Close_Time_Seconds'] / 60 / 60 / 24

In [111]:
median_close_time

DATETIMEINIT  Close_Time_Seconds  Year  Close_Time_Days
0    2010-12-31            271677.0  2010         3.144410
1    2011-12-31            344876.0  2011         3.991620
2    2012-12-31            438460.0  2012         5.074769
3    2013-12-31            518555.0  2013         6.001794
4    2014-12-31            428056.0  2014         4.954352
5    2015-12-31            445098.0  2015         5.151597
6    2016-12-31            518028.0  2016         5.995694
7    2017-12-31            357021.0  2017         4.132187
8    2018-12-31            352086.0  2018         4.075069
9    2019-12-31            511923.0  2019         5.925035
10   2020-12-31            360122.5  2020         4.168084

### we only want a dataframe with 'Year' and Close_Time_Days'

In [112]:
median_close_time = median_close_time[['Year', 'Close_Time_Days']].copy()

### let us view our new dataframe

In [113]:
median_close_time

Year  Close_Time_Days
0   2010         3.144410
1   2011         3.991620
2   2012         5.074769
3   2013         6.001794
4   2014         4.954352
5   2015         5.151597
6   2016         5.995694
7   2017         4.132187
8   2018         4.075069
9   2019         5.925035
10  2020         4.168084

# I am now going to merge the two dataframes
Using pd.merge, I will merge annual cases and median-close-time to create one dataframe called annual_summary using outer joint.

First, lets view both dataframes

In [116]:
annual_cases

Case_Count  Year
0        31504  2010
1        33915  2011
2        36560  2012
3        40243  2013
4        45399  2014
5        48990  2015
6        53969  2016
7        52741  2017
8        49873  2018
9        76886  2019
10       70774  2020

In [117]:
median_close_time

Year  Close_Time_Days
0   2010         3.144410
1   2011         3.991620
2   2012         5.074769
3   2013         6.001794
4   2014         4.954352
5   2015         5.151597
6   2016         5.995694
7   2017         4.132187
8   2018         4.075069
9   2019         5.925035
10  2020         4.168084

In [118]:
annual_summary = pd.merge(
    median_close_time,
    annual_cases,
    on='Year',
    how='outer',
    validate='1:1'
)

### Now let's view our newest dataframe

In [119]:
annual_summary

Year  Close_Time_Days  Case_Count
0   2010         3.144410       31504
1   2011         3.991620       33915
2   2012         5.074769       36560
3   2013         6.001794       40243
4   2014         4.954352       45399
5   2015         5.151597       48990
6   2016         5.995694       53969
7   2017         4.132187       52741
8   2018         4.075069       49873
9   2019         5.925035       76886
10  2020         4.168084       70774